addons/udes_stock_refactoring/models/stock_move.py

```diff
@@ -0,0 +1,334 @@
+    "confirmed": "confirm",
+    "waiting": "confirm",
+    "assigned": "assign",
+    "partially_available": "assign",
+    "done": "validate",
+}
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def _get_refactor_class(self, picking_type, stage):
+        """
+        Return the relevant refactor class from the provided 
+        picking type and stage, if applicable.
+        """
+        stage_post_action_dict = {
+            "confirm": picking_type.u_post_confirm_action,
+            "assign": picking_type.u_post_assign_action,
+            "validate": picking_type.u_post_validate_action,
+        }
+
+        refactor_action = stage_post_action_dict.get(stage)
+
+        if refactor_action:
+            try:
+                refactor_class = REFACTOR_REGISTRY[refactor_action]
+                return refactor_class(self.env)
+            except KeyError:
+                raise ValueError(_("Refactor action '%s' could not be found." % refactor_action))
+
+        return refactor_action
+
+    def compute_grouping_key(self):
+        """Compute the grouping key from the picking type move key format"""
+        # TODO MTC: This can be refactored and abstracted at some point with
+        # this with the equivalent for move_line_key
+        # we need to think of how we want to do it
+
+        # TODO: Look into making this computed field stored so it only needs 
+        # to recalculate when the picking type move key format is updated
+        # and check for improvements to performance.
+
+        # The environment must include {'compute_key': True}
+        # to allow the keys to be computed.
+        if not self.env.context.get("compute_key", False):
+            return
+        for move in self:
+            move_vals = {
+                field_name: move[field_name]
+                for field_name in move._fields.keys()
+                if field_name != "u_grouping_key"
+            }
+
+            format_str = move.picking_id.picking_type_id.u_move_key_format
+
+            if format_str:
+                move.u_grouping_key = format_str.format(**move_vals)
+            else:
+                move.u_grouping_key = None
+
+    u_grouping_key = fields.Char("Key", compute="compute_grouping_key")
+
+    def action_refactor(self):
+        """
+        Refactor all the moves in self. May result in the moves being changed
+        and/or their associated pickings being deleted.
+        """
+        self._action_refactor()
+        return True
+
+    def _action_refactor(self, stage=None):
+        """
+        Refactor moves in self if any exist.
+        :param stage: One of confirm|assign|done, if set, filters the moves
+            which will be refactored to only the state(s) that match:
+                - 'confirm': confirmed, waiting
+                - 'assign': assigned, partially_available
+                - 'done': done
+
+           Methods doing a refactor are expected to take a single recordset of
```

Comment:
Nitpic: indentation of this bit of the docstring

Reply:
Have now fixed this.